### ISSUE TITLE
Give users ability to define whether a setting should be shared

### ISSUE THREAD

**Issue by mirfilip** [ovity/octotree#396]
Currently, you override [WebAPI storage](https://github.com/buunguyen/octotree/blob/master/src/config/chrome/overrides.js#L2) to apply a strategy whether to use `chrome.storage` or `localStorage` for storing extension settings.

This is a cool idea. Yet, the key is stored in the `chrome.storage` only if it ends with `.shared` [ref](https://github.com/buunguyen/octotree/blob/master/src/config/chrome/overrides.js#L8).

It has two consequences for Chrome users:

- it is less secure because `WebAPI storage` is accessible to a website. This matters if you store a token. Yes, it's less likely to have an XSS on github or github enterprise so that local storage can leak, but still.
- it does not facilitate sync functions of `chrome.storage`

There are more but these two are most important. I actually came across this because my token wasn't synced between machines.

Currently, [only two options as marked as shared](https://github.com/buunguyen/octotree/blob/master/src/constants.js#L5). I may not know whole picture, but is there any reason for not using that?

As for the UI,  that could be done by "shared" checkbox near the particular field in settings or just one checkbox to sync all the settings with Chrome Sync (that's the way many extensions work).

**Comment by buunguyen** [comment]
Most are not shared because Octotree can be used in enterprise deployment as well, each might need to be configured differently. This is a must-do for the access token, which differs for each site. We can get around this by providing the UI for user to enterprise access token to each site separately.

As to sync, it was discussed here: https://github.com/buunguyen/octotree/issues/183, I think the conclusion is similar to what you propose.

Would you mind working on a PR?

**Comment by buunguyen** [comment]
If all settings are sync for an account, it will also address this issue https://github.com/buunguyen/octotree/issues/347

**Comment by mirfilip** [comment]
I could address it sometime this month, if such a long schedule suits you.

Thinking about the UI, providing much granularity for syncing choice would clutter it. To start with, I would opt for "sync all" checkbox (default off) **only for Chrome**. Later, I could check for syncing options in other browser (I know FF has one) and how to programmatically check if that's supported in the browser.

What I propose above would tackle all the issues you listed.

**Comment by buunguyen** [comment]
Sounds good. If you could share a mockup of what you plan to implement before actually implementing it, would be great to make sure we're on the same page.

**Comment by mirfilip** [comment]
Yes, I will.

One more thing. I don't fully understand below

> Most are not shared because Octotree can be used in enterprise deployment as well, each might need to be configured differently. This is a must-do for the access token, which differs for each site. We can get around this by providing the UI for user to enter access token to each site separately.

`chrome.storage` syncing is done on a Chrome profile level. There is a thing, that a token seems to be kept per domain/site. E.g. I set up a token for the Github Enterprise but not for Github. Of course, that should be kept on per-domain/site basis. In above, do you mean 

> We can get around this by providing the UI for user to enter access token to each **domain** separately.

not

>We can get around this by providing the UI for user to enter access token to each site separately.

right?

If I miss anything, please elaborate.

**Comment by buunguyen** [comment]
Currently most settings are stored in local storage, thus per-site basis. Token is such setting. So when users is on GitHub, they can input one token, when in another site, they can supply another token. The `.shared` settings are stored in `chrome.storage.local` ("local" here means "not sync'ed", not "local storage"). So these shared settings will be available for Octotree regardless of sites the user is on.

For token to use chrome storage (local or sync), it has to be specific for the site without the mechanism of local storage. One thing to do this is to explicitly associate a site and a token. I imagine the design would look like this:

[github]   ->  [enter access token]  [add site]

When click "add site", will add another line:

[enter site url]   ->   [enter access token]

So basically the site setting is not instead of a list of URLs, it will be a dictionary (url -> token). Depending on the current site the user is on, we can get the associated token to use.

Let me know if it makes sense.

**Comment by marsjaninzmarsa** [comment]
Storing per-site settings as simple object inside `chrome.storage.sync` should work just fine:
```js
{
	"github.com": {
		token: "foo",
		biz: "buzz"
	},
	"example.com": {
		token: null,
		option: true
	}
}
```

**Comment by timofonic** [comment]
@mirfilip @buunguyen @marsjaninzmarsa Any news?

**Comment by buunguyen** [comment]
Move discussion here https://github.com/ovity/octotree/issues/763
